theoriq/api/v1alpha1/execute.py

```diff
@@ -0,0 +1,92 @@
+"""
+execute.py
+
+Types and functions used by an Agent when executing a Theoriq request
+"""
+
+from __future__ import annotations
+
+from typing import Callable, List, Sequence
+
+from theoriq.agent import Agent
+from theoriq.biscuit import RequestBiscuit, TheoriqBudget, TheoriqRequest
+from theoriq.dialog import Dialog, DialogItem, ItemBlock
+from theoriq.types import Metric
+
+from ..common import ExecuteContextBase, ExecuteResponse
+from .protocol.protocol_client import ProtocolClient
+from .schemas.request import ExecuteRequestBody
+
+
+class ExecuteContext(ExecuteContextBase):
+    """
+    Represents the context for executing a request, managing interactions with the agent and protocol client.
+    """
+
+    def __init__(self, agent: Agent, protocol_client: ProtocolClient, request_biscuit: RequestBiscuit) -> None:
+        """
+        Initializes an ExecuteContext instance.
+
+        Args:
+            agent (Agent): The agent responsible for handling the execution.
+            protocol_client (ProtocolClient): The client responsible for communicating with the protocol.
+            request_biscuit (RequestBiscuit): The biscuit associated with the request, containing metadata and permissions.
+        """
+        super().__init__(agent, request_biscuit)
+        self._protocol_client = protocol_client
+
+    def send_event(self, message: str) -> None:
+        """
+        Sends an event message via the protocol client.
+
+        Args:
+            message (str): The message to send as an event.
+        """
+        self._protocol_client.post_event(request_biscuit=self._request_biscuit, message=message)
+
+    def send_metrics(self, metrics: List[Metric]):
+        """
+        Sends agent metrics via the protocol client.
+
+        Args:
+            metrics (List[MetricRequest]): The list of metrics to send.
+        """
+        self._protocol_client.post_metrics(request_biscuit=self._request_biscuit, metrics=metrics)
+
+    def send_metric(self, metric: Metric):
+        """
+        Sends agent metrics via the protocol client.
+
+        Args:
+            metric (MetricRequest): The metric to send.
+        """
+        self._protocol_client.post_metrics(request_biscuit=self._request_biscuit, metrics=[metric])
+
+    def send_request(self, blocks: Sequence[ItemBlock], budget: TheoriqBudget, to_addr: str) -> ExecuteResponse:
```

Comment:
Probably for another PR, but there might be a need to have a the ability to send a request with `Dialog` in addition to sending a request with `Blocs`